### ISSUE TITLE
Variable Length Trials not working? Or are they?

### ISSUE THREAD

**Issue by robintibor** [braindecode/braindecode#222]
It seems at the moment cropped scoring does not work for variable length trials? Although I have not really clearly confirmed that yet.
Here is an example that fails (unfortunately not runnable for others as it relies on local data on our cluster):
https://gist.github.com/robintibor/343b7b54fea558a7c73ea93d167490a4

Anyways @sliwy I had some problem to understand the code here:
https://github.com/braindecode/braindecode/blob/41d6fc4479b8a3b7db443648d288e47981be29b1/braindecode/training/scoring.py#L178-L182

Looks to me like this could never work for variable length trials? Still for some older version the above code apparently ran? Next steps here might be to create a simpler example with artificial variable length trial dataset (we would need it anyways).. Or also run TUHAbnormal that also actually shouldn't work I guess? I can also look at it more next week just since I saw you active @sliwy  thought I put it public here :)

**Comment by sliwy** [comment]
I agree that it should not work at the moment. The solution might to pad each element of `trial_preds` with `np.nan` to be the same length. Then, each time we compute average over time we should use `np.nanmean` to drop those values. What do you think? @robintibor  It would be great if we can have a test scenario with variable trial length which should work in the end.

**Comment by robintibor** [comment]
The thing is as far as I understand the code we just forward the preds to skorch and have to rely on their logic to work, no idea what they do with regard to nanmean... have to look inside a bit again

**Comment by robintibor** [comment]
It seems in this commit we removed the averaging of the predictions in the scoring function (see towards bottom): 
https://github.com/braindecode/braindecode/commit/bd85babeeb9d90848d9e938b8f3946be2649cce6
This breaks some use cases, actually then it's also not really a "cropped trial" rather a "cropped crop" scoring function right?

**Comment by robintibor** [comment]
Probably here it's explained what is supposed to happen: https://github.com/braindecode/braindecode/pull/171 Maybe this you meant with the nanmean there?

**Comment by robintibor** [comment]
But regarding nan padding and nanmean... not so sure if it is good idea, seems make everything even harder to understand and also will hide errors when you actually do have nans inside.... wonder if there is a cleaner solution

**Comment by sliwy** [comment]
@robintibor I think we can do this without nans. It's good you reminded me about this PR. The quickest solution would be in my opinion to rollback EpochScoring to this version:
```python 
            y_preds_per_trial = np.array(
                [np.mean(p, axis=1) for p in trial_preds]
            )
```
and change `predict_proba` to something hacky like this:
```python
def predict_proba(self, X):
        y_pred = super().predict_proba(X)
        if self.cropped and len(y_pred.shape) == 3:
            return y_pred.mean(axis=-1)
        else:
            return y_pred
```
I tried in my fork https://github.com/sliwy/braindecode/tree/variable_trial_length, technically it works (at least `test_cropped_decoding.py`) but it gives different results than previously. You can try to find the problem there.  I think that it should be the quickest but hacky solution.

**Comment by sliwy** [comment]
@robintibor take a look at #224 . can you check it with your problem where you have variable trial length? From my understanding it should work now. Quite hacky in the end but we can think about improvements if it works :)

I worry that it may not work yet for prediction outside training---needs to be tested more but I don't have more time, so just leaving it here in case you want to use it.

**Comment by robintibor** [comment]
So is solved now with #224 , great.
